### PR TITLE
client: Handle absolute path as redirect location in HTTP client

### PR DESCRIPTION
**Pull request by anlambert**
Some git servers can redirect a repository URL using an absolute-path URI reference as value of the `Location` HTTP header as allowed by [RFC 7231 Section 7.1.2](https://www.rfc-editor.org/rfc/rfc7231#section-7.1.2).

```console
$ curl -I https://codeberg.org/ashwinvis/radicale-sh.git
HTTP/2 307 
cache-control: no-store, no-transform
content-type: text/html; charset=utf-8
location: /ashwinvis/radicale-auth-sh
set-cookie: i_like_gitea=7b9a4d2494e520a3; Path=/; HttpOnly; SameSite=Lax; Secure; SameSite=Lax
set-cookie: _csrf=rcOoJeB1NS-JKRoc8ve8til2uVs6MTY3ODc0MDA0NTE0MzgwMzI1MQ; Path=/; Expires=Tue, 14 Mar 2023 20:40:45 GMT; HttpOnly; SameSite=Lax; Secure; SameSite=Lax
set-cookie: macaron_flash=; Path=/; Max-Age=0; HttpOnly; SameSite=Lax; Secure; SameSite=Lax
date: Mon, 13 Mar 2023 20:40:45 GMT
strict-transport-security: max-age=63072000; includeSubDomains; preload
permissions-policy: interest-cohort=()
x-frame-options: sameorigin
x-content-type-options: nosniff
content-security-policy-report-only: default-src data: 'self' https://*.codeberg.org https://codeberg.org; script-src 'self' https://*.codeberg.org https://codeberg.org; style-src data: 'self' 'unsafe-inline' https://*.codeberg.org https://codeberg.org; img-src *; media-src *; object-src 'none'; report-uri https://codeberg.org/.well-known/csp-report

$ git clone https://codeberg.org/ashwinvis/radicale-sh.git
Cloning into 'radicale-sh'...
warning: redirecting to https://codeberg.org/ashwinvis/radicale-auth-sh/
remote: Enumerating objects: 54, done.
remote: Total 54 (delta 0), reused 0 (delta 0), pack-reused 54
Receiving objects: 100% (54/54), 11.06 KiB | 707.00 KiB/s, done.
Resolving deltas: 100% (20/20), done.
```

That edge case was causing `dulwich` to raise `urllib3.exceptions.LocationValueError` when attempting to clone such repository as the redirection URL was missing the scheme and the net location.

```console
$ swh loader run git https://codeberg.org/ashwinvis/radicale-sh.git
WARNING:swh.loader.cli:No storage configuration detected, using an in-memory storage instead.
INFO:swh.loader.git.loader.GitLoader:Load origin 'https://codeberg.org/ashwinvis/radicale-sh.git' with type 'git'
INFO:urllib3.poolmanager:Redirecting https://codeberg.org/ashwinvis/radicale-sh.git/info/refs?service=git-upload-pack -> https://codeberg.org/ashwinvis/radicale-auth-sh/info/refs?service=git-upload-pack
ERROR:swh.loader.git.loader.GitLoader:Loading failure, updating to `failed` status
Traceback (most recent call last):
  File "/home/anlambert/swh/swh-environment/swh-loader-core/swh/loader/core/loader.py", line 417, in load
    more_data_to_fetch = self.fetch_data()
  File "/home/anlambert/swh/swh-environment/swh-loader-git/swh/loader/git/loader.py", line 319, in fetch_data
    fetch_info = self.fetch_pack_from_origin(
  File "/home/anlambert/swh/swh-environment/swh-loader-git/swh/loader/git/loader.py", line 237, in fetch_pack_from_origin
    pack_result = client.fetch_pack(
  File "/home/anlambert/dev/dulwich/dulwich/client.py", line 2113, in fetch_pack
    resp, read = self._smart_request(
  File "/home/anlambert/dev/dulwich/dulwich/client.py", line 1989, in _smart_request
    resp, read = self._http_request(url, headers, data)
  File "/home/anlambert/dev/dulwich/dulwich/client.py", line 2212, in _http_request
    resp = self.pool_manager.request(
  File "/home/anlambert/.virtualenvs/swh/lib/python3.9/site-packages/urllib3/request.py", line 78, in request
    return self.request_encode_body(
  File "/home/anlambert/.virtualenvs/swh/lib/python3.9/site-packages/urllib3/request.py", line 170, in request_encode_body
    return self.urlopen(method, url, **extra_kw)
  File "/home/anlambert/.virtualenvs/swh/lib/python3.9/site-packages/urllib3/poolmanager.py", line 365, in urlopen
    conn = self.connection_from_host(u.host, port=u.port, scheme=u.scheme)
  File "/home/anlambert/.virtualenvs/swh/lib/python3.9/site-packages/urllib3/poolmanager.py", line 237, in connection_from_host
    raise LocationValueError("No host specified.")
urllib3.exceptions.LocationValueError: No host specified.
{'status': 'failed'} for origin 'https://codeberg.org/ashwinvis/radicale-sh.git'
```

Fix the issue by restoring scheme and net location to the redirection URL using those from the original request URL.

### CODE REVIEW

File: dulwich/client.py [hunk 1]
```diff
@@ -1948,6 +1948,15 @@ def _discover_references(self, service, base_url):
                     % (url, resp.redirect_location, tail)
                 )
             base_url = resp.redirect_location[: -len(tail)]
+            base_url_parsed = urlparse(base_url)
```
Comment (jelmer) [inline dulwich/client.py:1951]: could this just use urljoin?

```suggestion
            base_url = urllib.parse.urljoin(url, base_url)
```

Comment (anlambert) [reply dulwich/client.py:1951]: This works for the edge case fixed in that PR but not for some classical  redirection URLs, see failing test below:
```
_________________________________________________________________________________ HttpGitClientTests.test_url_redirect_location __________________________________________________________________________________

self = <dulwich.tests.test_client.HttpGitClientTests testMethod=test_url_redirect_location>

    def test_url_redirect_location(self):
        from urllib3.response import HTTPResponse
    
        test_data = {
            "https://gitlab.com/inkscape/inkscape/": {
                "location": "https://gitlab.com/inkscape/inkscape.git/",
                "redirect_url": "https://gitlab.com/inkscape/inkscape.git/",
                "refs_data": (
                    b"001e# service=git-upload-pack\n00000032"
                    b"fb2bebf4919a011f0fd7cec085443d0031228e76 "
                    b"HEAD\n0000"
                ),
            },
            "https://github.com/jelmer/dulwich/": {
                "location": "https://github.com/jelmer/dulwich/",
                "redirect_url": "https://github.com/jelmer/dulwich/",
                "refs_data": (
                    b"001e# service=git-upload-pack\n00000032"
                    b"3ff25e09724aa4d86ea5bca7d5dd0399a3c8bfcf "
                    b"HEAD\n0000"
                ),
            },
            # check for absolute-path URI reference as location
            "https://codeberg.org/ashwinvis/radicale-sh.git/": {
                "location": "/ashwinvis/radicale-auth-sh/",
                "redirect_url": "https://codeberg.org/ashwinvis/radicale-auth-sh/",
                "refs_data": (
                    b"001e# service=git-upload-pack\n00000032"
                    b"470f8603768b608fc988675de2fae8f963c21158 "
                    b"HEAD\n0000"
                ),
            },
        }
    
        tail = "info/refs?service=git-upload-pack"
    
        # we need to mock urllib3.PoolManager as this test will fail
        # otherwise without an active internet connection
        class PoolManagerMock:
            def __init__(self):
                self.headers = {}
    
            def request(self, method, url, fields=None, headers=None, redirect=True, preload_content=True):
                base_url = url[: -len(tail)]
                redirect_base_url = test_data[base_url]["location"]
                redirect_url = redirect_base_url + tail
                headers = {
                    "Content-Type": "application/x-git-upload-pack-advertisement"
                }
                body = test_data[base_url]["refs_data"]
                # urllib3 handles automatic redirection by default
                status = 200
                request_url = redirect_url
                # simulate urllib3 behavior when redirect parameter is False
                if redirect is False:
                    request_url = url
                    if redirect_base_url != base_url:
                        body = b""
                        headers["location"] = test_data[base_url]["location"]
                        status = 301
    
                return HTTPResponse(
                    body=BytesIO(body),
                    headers=headers,
                    request_method=method,
                    request_url=request_url,
                    preload_content=preload_content,
                    status=status,
                )
    
        pool_manager = PoolManagerMock()
    
        for base_url in test_data.keys():
            # instantiate HttpGitClient with mocked pool manager
            c = HttpGitClient(base_url, pool_manager=pool_manager, config=None)
            # call method that detects url redirection
            _, _, processed_url = c._discover_references(b"git-upload-pack", base_url)
    
            # send the same request as the method above without redirection
            resp = c.pool_manager.request("GET", base_url + tail, redirect=False)
    
            # check expected behavior of urllib3
            redirect_location = resp.get_redirect_location()
    
            if resp.status == 200:
                self.assertFalse(redirect_location)
    
            if redirect_location:
                # check that url redirection has been correctly detected
>               self.assertEqual(processed_url, test_data[base_url]["redirect_url"])
E               AssertionError: 'https://gitlab.com/inkscape/inkscape/' != 'https://gitlab.com/inkscape/inkscape.git/'
E               - https://gitlab.com/inkscape/inkscape/
E               + https://gitlab.com/inkscape/inkscape.git/
E               ?                                     ++++

dulwich/tests/test_client.py:1129: AssertionError
```

Comment (anlambert) [reply dulwich/client.py:1951]: In fact it does not work either for absolute path URI, previous test was failing at first error, I commented out some tests data to also test with it.
```
_________________________________________________________________________________ HttpGitClientTests.test_url_redirect_location __________________________________________________________________________________

self = <dulwich.tests.test_client.HttpGitClientTests testMethod=test_url_redirect_location>

    def test_url_redirect_location(self):
        from urllib3.response import HTTPResponse
    
        test_data = {
            # "https://gitlab.com/inkscape/inkscape/": {
            #     "location": "https://gitlab.com/inkscape/inkscape.git/",
            #     "redirect_url": "https://gitlab.com/inkscape/inkscape.git/",
            #     "refs_data": (
            #         b"001e# service=git-upload-pack\n00000032"
            #         b"fb2bebf4919a011f0fd7cec085443d0031228e76 "
            #         b"HEAD\n0000"
            #     ),
            # },
            "https://github.com/jelmer/dulwich/": {
                "location": "https://github.com/jelmer/dulwich/",
                "redirect_url": "https://github.com/jelmer/dulwich/",
                "refs_data": (
                    b"001e# service=git-upload-pack\n00000032"
                    b"3ff25e09724aa4d86ea5bca7d5dd0399a3c8bfcf "
                    b"HEAD\n0000"
                ),
            },
            # check for absolute-path URI reference as location
            "https://codeberg.org/ashwinvis/radicale-sh.git/": {
                "location": "/ashwinvis/radicale-auth-sh/",
                "redirect_url": "https://codeberg.org/ashwinvis/radicale-auth-sh/",
                "refs_data": (
                    b"001e# service=git-upload-pack\n00000032"
                    b"470f8603768b608fc988675de2fae8f963c21158 "
                    b"HEAD\n0000"
                ),
            },
        }
    
        tail = "info/refs?service=git-upload-pack"
    
        # we need to mock urllib3.PoolManager as this test will fail
        # otherwise without an active internet connection
        class PoolManagerMock:
            def __init__(self):
                self.headers = {}
    
            def request(self, method, url, fields=None, headers=None, redirect=True, preload_content=True):
                base_url = url[: -len(tail)]
                redirect_base_url = test_data[base_url]["location"]
                redirect_url = redirect_base_url + tail
                headers = {
                    "Content-Type": "application/x-git-upload-pack-advertisement"
                }
                body = test_data[base_url]["refs_data"]
                # urllib3 handles automatic redirection by default
                status = 200
                request_url = redirect_url
                # simulate urllib3 behavior when redirect parameter is False
                if redirect is False:
                    request_url = url
                    if redirect_base_url != base_url:
                        body = b""
                        headers["location"] = test_data[base_url]["location"]
                        status = 301
    
                return HTTPResponse(
                    body=BytesIO(body),
                    headers=headers,
                    request_method=method,
                    request_url=request_url,
                    preload_content=preload_content,
                    status=status,
                )
    
        pool_manager = PoolManagerMock()
    
        for base_url in test_data.keys():
            # instantiate HttpGitClient with mocked pool manager
            c = HttpGitClient(base_url, pool_manager=pool_manager, config=None)
            # call method that detects url redirection
            _, _, processed_url = c._discover_references(b"git-upload-pack", base_url)
    
            # send the same request as the method above without redirection
            resp = c.pool_manager.request("GET", base_url + tail, redirect=False)
    
            # check expected behavior of urllib3
            redirect_location = resp.get_redirect_location()
    
            if resp.status == 200:
                self.assertFalse(redirect_location)
    
            if redirect_location:
                # check that url redirection has been correctly detected
>               self.assertEqual(processed_url, test_data[base_url]["redirect_url"])
E               AssertionError: 'https://codeberg.org/ashwinvis/radicale-sh.git/' != 'https://codeberg.org/ashwinvis/radicale-auth-sh/'
E               - https://codeberg.org/ashwinvis/radicale-sh.git/
E               ?                                           ----
E               + https://codeberg.org/ashwinvis/radicale-auth-sh/
E               ?
```

Comment (jelmer) [reply dulwich/client.py:1951]: You'd still want to do the manipulation on line 1950, i.e.:

```
diff --git a/dulwich/client.py b/dulwich/client.py
index 128da1beb..dbd46653a 100644
--- a/dulwich/client.py
+++ b/dulwich/client.py
@@ -1947,7 +1947,7 @@ class AbstractHttpGitClient(GitClient):
                     "Redirected from URL %s to URL %s without %s"
                     % (url, resp.redirect_location, tail)
                 )
-            base_url = resp.redirect_location[: -len(tail)]
+            base_url = urljoin(url, resp.redirect_location[: -len(tail)])

         try:
             self.dumb = (
```

Comment (anlambert) [reply dulwich/client.py:1951]: Indeed, bad copy paste from my side. Much simpler indeed, will update.

**Review by jelmer — APPROVED** [review]

**Comment by jelmer** [comment]
@efiop 0.21.5 is out now

**Comment by efiop** [comment]
@jelmer Thank you so much! 🙏
